crates/bevy_ecs/src/reflect/entity_commands.rs

```diff
@@ -0,0 +1,318 @@
+        type_registry: TypeRegistry,
+    ) -> &mut Self;
+
+    /// Removes the component of the same type as the supplied boxed reflect component from the entity
+    /// using the reflection data in the supplied [`TypeRegistry`]. Does nothing if the entity does not
+    /// have a component of the same type, if the [`TypeRegistry`] does not contain the reflection data
+    /// for the given component, or if the entity does not exist.
+    ///
+    /// # Note
+    /// Prefer to use the typed [`EntityCommands::remove`] unless you have good reason to use reflection.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    ///
+    /// # use bevy_ecs::prelude::*;
+    /// # use bevy_ecs::reflect::EntityCommandsReflectExtension;
+    /// # use bevy_reflect::{FromReflect, FromType, Reflect, TypeRegistry};
+    ///
+    /// # #[derive(Resource)]
+    /// # struct TypeRegistryResource{
+    /// #     type_registry: TypeRegistry,
+    /// # }
+    /// // A resource that can hold any component that implements reflect as a boxed reflect component
+    /// #[derive(Resource)]
+    /// struct SpecialComponentHolder{
+    ///     entity: Entity,
+    ///     component: Box<dyn Reflect>,
+    /// }
+    /// #[derive(Component, Reflect, FromReflect, Default)]
+    /// #[reflect(Component)]
+    /// struct ComponentA(u32);
+    /// #[derive(Component, Reflect, FromReflect, Default)]
+    /// #[reflect(Component)]
+    /// struct ComponentB(String);
+    ///
+    /// fn remove_reflected_component(
+    ///     mut commands: Commands,
+    ///     mut type_registry: ResMut<TypeRegistryResource>,
+    ///     special_component_holder: Res<SpecialComponentHolder>
+    ///     ) {
+    ///     #
+    ///     # type_registry.type_registry.register::<ComponentA>();
+    ///     # let mut registration = type_registry
+    ///     #     .type_registry
+    ///     #     .get_mut(std::any::TypeId::of::<ComponentA>())
+    ///     #     .unwrap();
+    ///     # registration.insert(<ReflectComponent as FromType<ComponentA>>::from_type());
+    ///     #
+    ///     # type_registry.type_registry.register::<ComponentB>();
+    ///     # let mut registration = type_registry
+    ///     #     .type_registry
+    ///     #     .get_mut(std::any::TypeId::of::<ComponentB>())
+    ///     #     .unwrap();
+    ///     # registration.insert(<ReflectComponent as FromType<ComponentB>>::from_type());
+    ///     // SpecialComponentHolder can hold any boxed reflect component. In this case either
+    ///     // ComponentA or ComponentB. No matter which component is in the resource though,
+    ///     // we can attempt to remove any component of that same type from an entity.
+    ///     commands.entity(special_component_holder.entity)
+    ///         .remove_reflected(special_component_holder.component.clone_value(), type_registry.type_registry.clone());
+    /// }
+    ///
+    /// ```
+    fn remove_reflected(
+        &mut self,
+        component: Box<dyn Reflect>,
```

Comment:
Hm, should this just take the type name? We don't really need any other data in order to remove the component.

Reply:
It should really be a `TypeId` . The example could show usage of `.remove_reflected(component.type_id())`.

And alternative is to define an `enum ReflectIdentifier` with a `TypeId` and `String` variant, and this function could accept a `impl Into<ReflectIdentifier>`

Reply:
Yeah `TypeId` would be nice, but we have to use the type name in order to support dynamic types (which will be very common in user code)

Reply:
Right but the concern here is protecting users unfamiliar with reflection. If they did this, they'd be confused why their component wasn't removed:
```rust
let data = component.clone_value();
// ...
commands.entity(id).remove_reflected(data.type_id(), registry.clone());
```

The only proper way to get the `TypeId` is to get it from the `TypeInfo`. But that's often less performant than just getting the type name.

Reply:
I guess this would be `data.as_any().type_id()` here. I was convinced there was a `Reflect::type_id` method that always returned the represented type's `TypeId`!  Seems it is not the case. In this case I agree we should only accept a `String`.

Reply:
Is this what the consensus is to change it too?
```rust
    fn remove_reflected(
        &mut self,
        component_type_name: String,
        type_registry: TypeRegistry,
    ) -> &mut Self;

```
and then calling it would look like?

```rust
 commands.entity(prefab.entity).remove_reflected(prefab.component.type_name().into(), type_registry.type_registry.clone());
```

If so is the example/naming enough of a notice or should there be some sort of call out in the documentation pointing towards the type_name parameter? I considered it initially but I was slightly worried that just accepting a String might be somewhat confusing for someone.

Reply:
Yeah, the example should be enough. We don't have a `TypeName` newtype, so it should be `String`. `component_type_name: impl Into<String>` is also a possibility, in order to avoid the `.into()` at the call site. But it also makes the type signature a bit harder to read. It's up to you. I've no preference.